jetty-jaspi/src/main/java/org/eclipse/jetty/security/jaspi/JaspiAuthenticator.java

```diff
@@ -15,72 +15,119 @@
 import org.eclipse.jetty.server.Authentication.User;
 import org.eclipse.jetty.server.UserIdentity;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+
+import static org.eclipse.jetty.security.jaspi.JaspiAuthenticatorFactory.MESSAGE_LAYER;
 
 /**
- * @version $Rev: 4793 $ $Date: 2009-03-19 00:00:01 +0100 (Thu, 19 Mar 2009) $
+ * Implementation of Jetty {@link LoginAuthenticator} that is a bridge from Jakarta Authentication to Jetty Security.
  */
+@SuppressWarnings({"rawtypes", "unchecked"})
 public class JaspiAuthenticator extends LoginAuthenticator
 {
-    private static final Logger LOG = LoggerFactory.getLogger(JaspiAuthenticator.class.getName());
-
-    private final ServerAuthConfig _authConfig;
-
-    private final Map _authProperties;
-
-    private final ServletCallbackHandler _callbackHandler;
-
     private final Subject _serviceSubject;
-
+    private final String _appContext;
     private final boolean _allowLazyAuthentication;
+    private final AuthConfigFactory _authConfigFactory = AuthConfigFactory.getFactory();
+    private Map _authProperties;
+    private IdentityService _identityService;
+    private ServletCallbackHandler _callbackHandler;
+    private ServerAuthConfig _authConfig;
 
-    private final IdentityService _identityService;
+    public JaspiAuthenticator(Subject serviceSubject, String appContext, boolean allowLazyAuthentication)
+    {
+        _serviceSubject = serviceSubject;
+        _appContext = appContext;
+        _allowLazyAuthentication = allowLazyAuthentication;
+    }
 
-    public JaspiAuthenticator(ServerAuthConfig authConfig, Map authProperties, ServletCallbackHandler callbackHandler, Subject serviceSubject,
-                              boolean allowLazyAuthentication, IdentityService identityService)
+    @Deprecated
+    public JaspiAuthenticator(ServerAuthConfig authConfig, Map authProperties, ServletCallbackHandler callbackHandler, Subject serviceSubject, boolean allowLazyAuthentication, IdentityService identityService)
     {
-        // TODO maybe pass this in via setConfiguration ?
         if (callbackHandler == null)
             throw new NullPointerException("No CallbackHandler");
         if (authConfig == null)
             throw new NullPointerException("No AuthConfig");
-        this._authConfig = authConfig;
         this._authProperties = authProperties;
         this._callbackHandler = callbackHandler;
         this._serviceSubject = serviceSubject;
         this._allowLazyAuthentication = allowLazyAuthentication;
         this._identityService = identityService;
+        this._appContext = null;
+        this._authConfig = authConfig;
     }
 
     @Override
     public void setConfiguration(AuthConfiguration configuration)
     {
+        LoginService loginService = configuration.getLoginService();
+        if (loginService == null)
+        {
+            // Add an empty login service so we can use JASPI without tying into Jetty auth mechanisms.
```

Comment:
I don't see how having a fake LoginService allows a webapp to do it's own jaspi authentication?

I think if there's no LoginService, isn't this an error?